runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DockerJobBundleFactory.java

```diff
@@ -69,6 +69,14 @@
 public class DockerJobBundleFactory implements JobBundleFactory {
   private static final Logger LOG = LoggerFactory.getLogger(DockerJobBundleFactory.class);
 
+  /** Factory that creates {@link JobBundleFactory} for the given {@link JobInfo}. */
+  public interface JobBundleFactoryFactory {
+    DockerJobBundleFactory create(JobInfo jobInfo) throws Exception;
+  }
+  // TODO: a hacky way to override the factory for testing.
+  // Should be replaced with mechanism that let's users configure their own factory
+  public static JobBundleFactoryFactory factory = (jobInfo) -> new DockerJobBundleFactory(jobInfo);
```

Comment:
Shall we make it final or remove it whole together. We can pass the right JobBundleFactory instance to avoid FactoryFactory.
This might need restructuring of the code a bit.

Reply:
It cannot be made final as it is used by the test. But in the future we might use a service loader to find the factory.